File: 230.二叉搜索树中第k小的元素.java
/*
 * @lc app=leetcode.cn id=230 lang=java
 *
 * [230] 二叉搜索树中第K小的元素
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // 二叉搜索树特点是左边比根节点小，右边比根节点大
    // 中序遍历的结果就是把二叉树从小到大遍历一下
    // 找到中序遍历的第k个元素就是结果

    // 记录结果
    int res = 0;
    // 记录当前的序号
    int index = 0;

    public int kthSmallest(TreeNode root, int k) {
        traverse(root, k);
        return res;
    }

    private void traverse(TreeNode node, int k) {
        if (node == null) {
            return;
        }
    
        traverse(node.left, k);
        // 中序遍历
        index++;
        if (index == k) {
            res = node.val;
            return;
        }
        traverse(node.right, k);
    }
 }
// @lc code=end

File: 230.二叉搜索树中第k小的元素.js
/*
 * @lc app=leetcode.cn id=230 lang=javascript
 *
 * [230] 二叉搜索树中第K小的元素
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {

    let res = 0;
    let n = 0;

    let traversal = function(node, k) {
        if (!node) {
            return;
        }
        
        traversal(node.left, k);

        n++;
        if (n === k) {
            res = node.val;
            return;
        }

        traversal(node.right, k);
    }

    traversal(root, k);
    return res;

};
// @lc code=end

